error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/TypeMemberOrder.java

```diff
@@ -0,0 +1,260 @@
+    }
+
+    /* All members that can be moved or may lay between movable ones. */
+    ImmutableList<TypeMember> members =
+        tree.getMembers().stream()
+            .filter(
+                member -> !MoreASTHelpers.isGeneratedConstructor(member) && !isEnumerator(member))
+            .map(m -> new AutoValue_TypeMemberOrder_TypeMember(m, getPreferredOrdinal(m, state)))
+            .collect(toImmutableList());
+
+    /*
+      List of the sortable members' preferred ordinals, ordered by the member's position in the
+      original source.
+    */
+    ImmutableList<Integer> preferredOrdinals =
+        members.stream()
+            .filter(m -> m.preferredOrdinal().isPresent())
+            .map(m -> m.preferredOrdinal().orElseThrow(/* Unreachable due to preceding check. */ ))
+            .collect(toImmutableList());
+
+    if (Comparators.isInOrder(preferredOrdinals, naturalOrder())) {
+      return Description.NO_MATCH;
+    }
+
+    int bodyStartPos = getBodyStartPos(tree, state);
+    if (bodyStartPos == Position.NOPOS) {
+      /*
+       * We can't determine the type body's start position in the source code. This generally means
+       * that (part of) its code was generated. Even if the source code for a subset of its members
+       * is available, dealing with this edge case is not worth the trouble.
+       */
+      return Description.NO_MATCH;
+    }
+
+    return describeMatch(tree, sortTypeMembers(bodyStartPos, members, state));
+  }
+
+  /**
+   * Returns true if {@link Tree} is an enumerator of an enumerated type, false otherwise.
+   *
+   * @see com.sun.tools.javac.tree.Pretty#isEnumerator(JCTree)
+   * @see com.sun.tools.javac.code.Flags#ENUM
+   */
+  private static boolean isEnumerator(Tree tree) {
+    return tree instanceof JCVariableDecl variableDecl && (variableDecl.mods.flags & ENUM) != 0;
+  }
+
+  /**
+   * Returns the preferred ordinal of the given member, or empty if it's unmovable for any reason,
+   * including it lacking a preferred ordinal.
+   */
+  private Optional<Integer> getPreferredOrdinal(Tree tree, VisitorState state) {
+    if (isSuppressed(tree, state)) {
+      return Optional.empty();
+    }
+    return switch (tree.getKind()) {
+      case VARIABLE -> Optional.of(isStatic((VariableTree) tree) ? 1 : 2);
+      case BLOCK -> Optional.of(isStatic((BlockTree) tree) ? 3 : 4);
+      case METHOD -> Optional.of(isConstructor((MethodTree) tree) ? 5 : 6);
+      case CLASS, INTERFACE, ENUM -> Optional.of(7);
+        // TODO: Should we log unhandled kinds?
+      default -> Optional.empty();
+    };
+  }
+
+  /**
+   * Returns the start position of the body of the given type, in the case of enums, it returns the
+   * position that follows the enumerated type's enumerations.
+   */
+  private static int getBodyStartPos(ClassTree tree, VisitorState state) {
+    CharSequence sourceCode = state.getSourceCode();
+    int typeStart = ASTHelpers.getStartPosition(tree);
```

Comment:
Usually we call this `startPosition`, I think we can do the same here 🤔, not sure if we should make it specific to add "type"? Or am I missing a reason to do so?